### DISCUSSION TITLE
Assigning a unique request ID to all traces within a request's lifetime

### DISCUSSION THREAD

**Discussion by jdglaser** [tokio-rs/axum#2273] [Q&A] [answered]
Hello,

I am relatively new to Axum / Tower / Tracing, so I apologize if this is in the wrong place.

I am trying to add a trace layer to my Axum service that adds a custom `request_id` attribute. What I ultimately want is to be able to trace the life of a request from start to finish. I know that `TraceLayer::new_for_http()` provides some nice defaults for this (for example the start and end of the request is logged). However, I also want any traces that happen between the start and end of a request to have the same `request_id`. So anytime I call `tracing::info!` or `tracing::debug!` from within some other service, I want the same `request_id` to be match until the end of the request.

I think I have something like this working using the following customizations to the `TraceLayer` service.

```rust
// ...
.layer(
  TraceLayer::new_for_http()
      .make_span_with(|request: &Request<Body>| {
          let request_id = uuid::Uuid::new_v4();
          tracing::span!(
              Level::DEBUG,
              "request",
              method = tracing::field::display(request.method()),
              uri = tracing::field::display(request.uri()),
              version = tracing::field::debug(request.version()),
              request_id = tracing::field::display(request_id)
          )
      })
)
// ...
``` 

When testing this locally, it actually appears to be working.

```
2023-10-16T15:26:30.329196Z  INFO request{method=GET uri=/api/current_user version=HTTP/1.1 request_id=d292da95-3402-4ceb-8922-15ca0ffe6c97}: tower_http::trace::on_request: started processing request
2023-10-16T15:26:30.329330Z  INFO request{method=GET uri=/api/current_user version=HTTP/1.1 request_id=d292da95-3402-4ceb-8922-15ca0ffe6c97}: backend::service::auth_service: Authorizing user
2023-10-16T15:26:30.329439Z  INFO request{method=GET uri=/api/current_user version=HTTP/1.1 request_id=d292da95-3402-4ceb-8922-15ca0ffe6c97}: tower_http::trace::on_response: finished processing request latency=0 ms status=200
```

I can see the `tower_http` trace logging the start of the request, a trace from a custom auth service that I called from within my code, and the `tower_http` trace logging the end of the request all share the same `request_id`.

What I am unsure of is how this will hold up in a "production" deployment with many threads and requests happening at once. My understanding of how `tracing` works within a multithreaded context is a bit murky. Is it possible that multiple threads handling requests will end up causing a different `request_id` to be set for the same request?

I would like to handle this in the way that I currently have it setup, as it feels really clean, but maybe the safest way to do this is to set the `request_id` in a request extension and always reference that in a `request_id` record when calling a `trace` macro.

Hopefully, this all makes sense. I really appreciate any feedback or guidance on this!

**Comment by davidpdrsn** [comment, marked as answer]
> My understanding of how tracing works within a multithreaded context is a bit murky.

Where did you hear that?

I think what you wrote should work fine.

**Reply by jdglaser** [reply]
Thank you for the quick response!

I guess just more conceptually I don’t really understand how tracing actually manages to keep different spans self contained when running multiple threads. Obviously, since axum is running with tokio with async/await as well as multiple threads handling requests, I wasn’t sure how that would work.

But sounds like I don’t need to worry, thanks!

**Reply by davidpdrsn** [reply]
I don’t know exactly how tracing does it but it’s made by smart people who I trust 😅

**Comment by ttys3** [comment]
what you need is tracing. just use the trace_id

**Reply by jdglaser** [reply]
Thanks for the quick response! I will take a look at trace_id!

**Comment by samuela** [comment]
What's the recommended approach to doing this? It would be great to include this use case in `examples/` IMHO!
